File: gmd-java/src/main/java/gov/lanl/gmd/queries/Queries.java
package gov.lanl.gmd.queries;

import java.awt.geom.Point2D;
import java.awt.geom.Point2D.Double;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * A class containing useful static queries on the magneto database.
 */
public class Queries {

	enum CoordinateType {GEOGRAPHIC,MAGNETIC}
	enum MeasurementOption {
		MLON("mlon"),
		MLAT("mlat"),
		N("n"),
		E("e"),
		Z("z");
		private String value;	
		MeasurementOption(String value){
			this.value = value;
		}
		public String toString(){
			return value;
		}
	}

	/**
	 * A station record object.
	 */
	public static class Station{
		private String iaga;
		private double glon;
		private double glat;
		private double mlon;
		private double mlat;
		private String station_name;

		public Station(String iaga, double glon, double glat, double mlon,
				double mlat, String station_name) {
			this.iaga = iaga;
			this.glon = glon;
			this.glat = glat;
			this.mlon = mlon;
			this.mlat = mlat;
			this.station_name = station_name;
		}
		public String getIaga() {
			return iaga;
		}
		public double getGlon() {
			return glon;
		}
		public double getGlat() {
			return glat;
		}
		public double getMlon() {
			return mlon;
		}
		public double getMlat() {
			return mlat;
		}
		public String getStation_name() {
			return station_name;
		}
	}

	/**
	 * A measurement record object.
	 */
	public static class Measurement{
		private Timestamp date_utc;
		private String iaga;
		private double mlt;
		private double mlat;
		private double n;
		private double e;
		private double z;

		public Measurement(Timestamp date_utc, String iaga, double mlt,
				double mlat, double n, double e, double z) {
			this.date_utc = date_utc;
			this.iaga = iaga;
			this.mlt = mlt;
			this.mlat = mlat;
			this.n = n;
			this.e = e;
			this.z = z;
		}
		public Timestamp getDate_utc() {
			return date_utc;
		}
		public String getIaga() {
			return iaga;
		}
		public double getMlt() {
			return mlt;
		}
		public double getMlat() {
			return mlat;
		}
		public double getN() {
			return n;
		}
		public double getE() {
			return e;
		}
		public double getZ() {
			return z;
		}			
	}

	public static class MeasurementFilter{
		private MeasurementOption[] measurementOptions = null;		
		private String[] stationIDs = null;
		private Timestamp[] timeRange = null;
		private Point2D[] coordRange = null;
		private CoordinateType coordinateType = CoordinateType.GEOGRAPHIC;
		
		public MeasurementFilter() {}
		
		public MeasurementOption[] getMeasurementOptions() {
			return measurementOptions;
		}

		public void setMeasurementOptions(MeasurementOption[] measurementOptions) {
			this.measurementOptions = measurementOptions;
		}

		public String[] getStationIDs() {
			return stationIDs;
		}
		public void setStationIDs(String[] stationIDs) {
			this.stationIDs = stationIDs;
		}
		public Timestamp[] getTimeRange() {
			return timeRange;
		}
		public void setTimeRange(Timestamp initialTime, Timestamp finalTime) {
			this.timeRange = new Timestamp[2];
			this.timeRange[0] = initialTime;
			this.timeRange[1] = finalTime;
		}
		public Point2D[] getCoordRange() {
			return coordRange;
		}
		public void setCoordRange(double minlon, double minlat, double maxlon, double maxlat) {
			this.coordRange = new Point2D[2];
			this.coordRange[0] = new Point2D.Double(minlon,minlat);
			this.coordRange[1] = new Point2D.Double(maxlon,maxlat);
		}
		public CoordinateType getCoordinateType() {
			return coordinateType;
		}
		public void setCoordinateType(CoordinateType coordinateType) {
			this.coordinateType = coordinateType;
		}
	}

	/**
	 * Schema descriptions.
	 */
	private static final String[] tableNames = {"stations","measurements"};
	private static final String[] stationFieldNames = {"iaga","glon","glat","mlon","mlat","station_name"};
	private static final Class<?>[] stationFieldTypes = {String.class, Double.class, Double.class,
		Double.class, Double.class, String.class};
	private static final Map<String,Class<?>> stationFields = new HashMap<>();
	static{
		for(int i=0;i<stationFieldNames.length;i++){
			stationFields.put(stationFieldNames[i], stationFieldTypes[i]);
		}
	}
	private static final String[] measurementFieldNames = {"date_utc","iaga","mlt","mlat","n","e","z","collection"};
	private static final Class<?>[] measurementFieldTypes = {Timestamp.class, String.class, Double.class,
		Double.class, Double.class, Double.class, Double.class, String.class};
	private static final Map<String,Class<?>> measurementFields = new HashMap<>();
	static{
		for(int i=0;i<measurementFieldNames.length;i++){
			measurementFields.put(measurementFieldNames[i], measurementFieldTypes[i]);
		}		
	}

	/**
	 * Returns a list of table names.
	 * @return
	 */
	protected static String[] getTableNames(){
		return tableNames;
	}

	/**
	 * Returns a list of station field names.
	 * @return
	 */
	public static String[] getStationFieldNames(){
		return stationFieldNames;
	}
	
	/**
	 * Returns a dictionary of station field names and types.
	 * @return
	 */
	public static Map<String,Class<?>> getStationFields(){
		return stationFields;
	}

	/**
	 * Returns a list of Station field types.
	 * @return
	 */
	protected static Class<?>[] getStationFieldTypes(){
		return stationFieldTypes;
	}

	/**
	 * Returns a list of measurement field names.
	 * @return
	 */
	public static String[] getMeasurementFieldNames(){
		return measurementFieldNames;
	}
	
	/**
	 * Returns a dictionary of measurement field names and types.
	 * @return
	 */
	public static Map<String,Class<?>> getMeasurementFields(){
		return measurementFields;
	}

	/**
	 * Returns a list of measurement field types.
	 * @return
	 */
	protected static Class<?>[] getMeasurementFieldTypes(){
		return measurementFieldTypes;
	}
	
	

	private static ResultSet queryMagneto(Connection conn, String queryString){
		ResultSet r = null;
		try {
			Statement statement = conn.createStatement();
			r = statement.executeQuery(queryString);
		} catch (SQLException e) {
			e.printStackTrace();
		}
		return r;
	}

	/**
	 * Returns a map of all station objects.
	 * @param conn
	 * @return
	 */
	public static Map<String,Station> getStations(Connection conn){
		String queryString = "select * from magneto.stations order by iaga";
		ResultSet r = queryMagneto(conn, queryString);
		return getStations(r);
	}


	public static Map<String,Station> getStationsByGeoCoords(Connection conn,
			double glonmin, double glatmin, double glonmax, double glatmax){
		String queryString = "select * from magneto.stations where"+
				" glon >= "+glonmin +" and glon <= "+glonmax+
				" and glat >= "+glatmin+" and glat <= "+glatmax+ " order by iaga";
		ResultSet r = queryMagneto(conn, queryString);
		return getStations(r);
	}

	public static Map<String,Station> getStationsByMagCoords(Connection conn,
			double mlonmin, double mlatmin, double mlonmax, double mlatmax){
		String queryString = "select * from magneto.stations where"+
				" mlon >= "+mlonmin+" and mlon <= "+mlonmax+
				" mlat >= "+mlatmin+" and mlat <= "+mlatmax+ " order by iaga";
		ResultSet r = queryMagneto(conn, queryString);
		return getStations(r);
	}

	private static Map<String,Station> getStations(ResultSet r){
		Map<String,Station> stations = null;
		try {
			if(r.next()){
				stations = new HashMap<>();
				while(r.next()){
					String iaga = r.getString("iaga");
					double glon = r.getDouble("glon");
					double glat = r.getDouble("glat");
					double mlon = r.getDouble("mlon");
					double mlat = r.getDouble("mlat");
					String station_name = r.getString("station_name");
					stations.put(iaga, new Station(iaga,glon,glat,mlon,mlat,station_name));
				}
			}
		} catch (SQLException e) {
			e.printStackTrace();
		}
		return stations;
	}

	public static ResultSet filterMeasurements(Connection conn, MeasurementFilter filter){
		ResultSet r = null;
		String[] stationIDs = filter.getStationIDs();
		Timestamp[] timeRange = filter.getTimeRange();
		Point2D[] coordRange = filter.getCoordRange();
		CoordinateType coordinateType = filter.getCoordinateType();
		if(timeRange!=null){ // Some type of time filter is required to perform the query.
			List<String> clauses = new ArrayList<>();
			Set<String> stationIDSet = null;
			// Station ID filter selects specific stations by ID.
			if(stationIDs!=null){
				if(stationIDSet==null){
					stationIDSet = new HashSet<>();
				}
				for(String stationID : stationIDs){
					stationIDSet.add(stationID);
				}
			}
			// Coordinate range filter selects stations by location.
			if(coordRange!=null){
				if(stationIDSet==null){
					stationIDSet = new HashSet<>();
				}
				Map<String,Station> stations = null;
				switch(coordinateType){
				case GEOGRAPHIC:
					stations = Queries.getStationsByGeoCoords(conn,coordRange[0].getX(),coordRange[0].getY(),
							coordRange[1].getX(),coordRange[1].getY());
					break;
				case MAGNETIC:
					stations = Queries.getStationsByMagCoords(conn,coordRange[0].getX(),coordRange[0].getY(),
							coordRange[1].getX(),coordRange[1].getY());
					break;
				}
				if(stations!=null){
					for(String stationID : stations.keySet()){
						stationIDSet.add(stationID);
					}
				}
			}
			// If the station set is not null, or empty, create a stations clause.
			if(stationIDSet!=null){
				if(stationIDSet.size()>0){
					String[] stationArray = stationIDSet.toArray(new String[0]);
					StringBuilder stationsClause = new StringBuilder("iaga = any(array['"+stationArray[0]);
					for(int i=1;i<stationArray.length;i++){
						stationsClause.append("','" + stationArray[i]);
					}
					stationsClause.append("'])");
					clauses.add(stationsClause.toString());
				}
			}
			if(timeRange!=null){
				StringBuilder timeRangeClause = new StringBuilder();
				timeRangeClause.append("date_utc >= '" + TimeFormatter.formatTimestamp(timeRange[0]));
				timeRangeClause.append("' and date_utc <= '" + TimeFormatter.formatTimestamp(timeRange[1])+"'");
				clauses.add(timeRangeClause.toString());
			}
			// Build the query string from the clauses.
			MeasurementOption[] measurementOptions = filter.getMeasurementOptions();
			StringBuilder columnSelection = new StringBuilder("date_utc,iaga");
			if(measurementOptions!=null)
			{
				columnSelection.append(","+measurementOptions[0]);
				for(int i=1;i<measurementOptions.length;i++){
					columnSelection.append(","+measurementOptions[i]);
				}
				columnSelection.append(",collection");
			} else{
				columnSelection.append("*");
			}
			StringBuilder queryString = new StringBuilder("select "+columnSelection.toString()
					+ " from magneto.measurements where ");
			queryString.append(clauses.get(0));
			if(clauses.size()>1){
				queryString.append(" and "+clauses.get(1));
			}
			queryString.append(" order by date_utc;");
			// System.out.println(queryString.toString());
			r = queryMagneto(conn,queryString.toString());
		}
		return r;
	}


}

File: gmd-python/magneto_api/queries.py
from datetime import datetime
from sqlalchemy.ext.automap import automap_base
from sqlalchemy import and_


class SchemaDescription:
    table_names = ["stations", "measurements"]
    station_fields = {"iaga":"character","glon":"double","glat":"double","mlon":"double",\
                      "mlat":"double","station_name":"character"}
    measurement_fields = {"date_utc":"timestamp","iaga":"character","mlt":"double","mlat":"double",\
                          "n":"double","e":"double","z":"double","collection":"character"}

class CoordinateType:
    GEOGRAPHIC = 0
    MAGNETIC = 1

class MeasurementOption:
    MLON = "mlon"
    MLAT = "mlat"
    N = "n"
    E = "e"
    Z = "z"

class Coordinate:
    def __init__(self,x,y):
        self.x = x
        self.y = y

# Timestamp range is given in python datetime.datetime instances.
class MeasurementFilter:
    def __init__(self):
        self.measurement_options = []
        self.station_ids = []
        self.timestamp_range = []
        self.coordinate_range = []
        self.coordinate_type = CoordinateType.GEOGRAPHIC

    def add_measurement_option(self,measurement_option):
        self.measurement_options.append(measurement_option)

    def add_station_ids(self,station_ids):
        for id in station_ids:
            self.station_ids.append(id)

    def set_timestamp_range(self,min_timestamp,max_timestamp):
        self.timestamp_range = []
        self.timestamp_range.append(min_timestamp)
        self.timestamp_range.append(max_timestamp)

    def set_coordinate_range(self,mincoord,maxcoord):
        self.coordinate_range = []
        self.coordinate_range.append(mincoord)
        self.coordinate_range.append(maxcoord)

    def set_coordinate_type(self,coordtype):
        self.coordinate_type = coordtype


def get_station_field_names():
    return list(SchemaDescription.station_fields.keys())

def get_station_field_types():
    return list(SchemaDescription.station_fields.values())

def get_measurement_field_names():
    return list(SchemaDescription.measurement_fields.keys())

def get_measurement_field_types():
    return list(SchemaDescription.measurement_fields.values())

def query_magneto(dbconnector, queryString):
    conn = dbconnector.engine.connect()
    rs = conn.execute(queryString)
    return rs

def get_stations(dbconnector):
    Station =dbconnector.Station
    stations = {}
    for station in dbconnector.session.query(Station).order_by(Station.iaga):
        stations[station.iaga] = station
    return stations

def get_stations_by_geocoords(dbconnector,minlon,minlat,maxlon,maxlat):
    Station =dbconnector.Station
    stations = {}
    for station in dbconnector.session.query(Station)\
            .filter(and_(Station.glon>=minlon,Station.glat>=minlat,Station.glon<=maxlon,Station.glat<=maxlat))\
            .order_by(Station.iaga):
        station.glon = station.glon
        stations[station.iaga] = station
    return stations

def get_stations_by_magcoords(dbconnector,minlon,minlat,maxlon,maxlat):
    # TODO Do these longitudes require any modification?
    Station =dbconnector.Station
    stations = {}
    for station in dbconnector.session.query(Station)\
            .filter(and_(Station.mlon>=minlon,Station.mlat>=minlat,Station.mlon<=maxlon,Station.mlat<=maxlat))\
            .order_by(Station.iaga):
        stations[station.iaga] = station
    return stations

def filter_measurements(dbconnector,filter):
    measurement_options = filter.measurement_options
    station_ids = filter.station_ids
    timestamp_range = filter.timestamp_range
    coordinate_range = filter.coordinate_range
    coordinate_type = filter.coordinate_type
    rs = None
    # Only perform the query if the time range filter is present at least.
    if (len(timestamp_range)>0):

        # Station ID set.
        station_set = set()

        # Add existing station IDs to the set.
        for station_id in station_ids:
            station_set.add(station_id)

        stations = None
        # If the coordinate range is not empty, query for stations in the coordinate range.
        if len(coordinate_range)>0:
            minlon = coordinate_range[0].x
            minlat = coordinate_range[0].y
            maxlon = coordinate_range[1].x
            maxlat = coordinate_range[1].y
            if coordinate_type == CoordinateType.GEOGRAPHIC:
                # Use geographic coordinates.
                stations = get_stations_by_geocoords(dbconnector,minlon,minlat,maxlon,maxlat)
            if coordinate_type == CoordinateType.MAGNETIC:
                # Use magnetic coordinates.
                stations = get_stations_by_magcoords(dbconnector,minlon,minlat,maxlon,maxlat)
        # Add stations in the range to the set.
        if stations!=None:
            key_list = stations.keys()
            for key in key_list:
                station_set.add(key)

        # Create a string-based query,filter based on the station set.
        # Check for measurement (column) options.
        column_list= "*"
        if len(measurement_options)>0:
            column_list = measurement_options[0]
            for i in range(1,len(measurement_options)):
                column_list =  column_list + ","+measurement_options[i]
            column_list = "date_utc,iaga,"+column_list+",collection"

        query_string = "select "+column_list+" from magneto.measurements where "

        clauses = []
        # Create a string for the stations clause.
        station_id_list = list(station_set)
        # If there are stations to be named.
        stations_clause = None
        if len(station_id_list)>0:
            stations_clause = "iaga = any(array['"+station_id_list[0]
            for i in range(1,len(station_id_list)):
                stations_clause = stations_clause +"','" + station_id_list[i]
            stations_clause = stations_clause +"'])"
            clauses.append(stations_clause)

        # Create a time range clause.
        time_range_clause = None
        if len(timestamp_range)>0:
            time_range_clause = "date_utc >= '" + str(timestamp_range[0]) \
                    + "' and date_utc <= '" + str(timestamp_range[1]) + "'"
            clauses.append(time_range_clause)

        # Form the query string.
        # Perform the query and return the measurements.
        query_string = query_string + clauses[0]
        if len(clauses)>1:
            query_string = query_string + " and " + clauses[1]
        query_string = query_string + " order by date_utc"
        rs = query_magneto(dbconnector,query_string)
    return rs


